llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -312,22 +314,161 @@ class RegionEndEdges : public ScheduleDAGMutation {
 /// Here, these special SUnits get created from Region::top_fixed_instrs() or
 /// Region::bot_fixed_instrs(), and dependencies are created between "free" and
 /// "fixed" SUnits.
+///
+/// When considering Region::top_fixed_instrs, we chain read and write
+/// events across different cycles. Rather than tracking individual
+/// dependencies, we establish a timeline of the most recent events, which can
+/// also occur in the preceding pipelined loop. Based on this timeline, we
+/// implicitly detect all dependencies for each instruction and determine a safe
+/// distance from the beginning of the region (EntrySU). Then, we create a
+/// single edge from EntrySU to each instruction. This approach means that we
+/// only need to track one dependency for each instruction, which also includes
+/// the parent pipelined loop, thereby eliminating the need to reassess
+/// dependencies related to EntrySU in other mutations
 class EmitFixedSUnits : public ScheduleDAGMutation {
 public:
+  // The purpose of this class is to implement a mechanism that tracks the
+  // cycles in which each register is last written or read. This information
+  // allows us to identify safe distances—measured in cycle units—where
+  // instructions can be placed without conflicts. These safe distances take
+  // into account read-after-write (RAW), write-after-read (WAR), and
+  // write-after-write (WAW) dependencies among all unscheduled instructions and
+  // timed bundles. This includes both top-insert instructions and their
+  // predecessor instructions in pipelined loops.
+  struct RegisterEventTracker {
+    using fixed_iterator = MachineBasicBlock::iterator;
+    const InstrItineraryData *InstrItins;
+    const TargetRegisterInfo *TRI;
+    std::map<MCRegister, unsigned> RegisterToCycleDef;
+    std::map<MCRegister, unsigned> RegisterToCycleUse;
```

Comment:
What happens if a same register appears multiple times in different operands? Will we be always be "conservative in the right direction"?